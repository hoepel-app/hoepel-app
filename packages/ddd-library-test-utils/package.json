{
  "name": "@hoepel.app/ddd-library-test-utils",
  "version": "0.3.113",
  "main": "dist/index.js",
  "license": "GPL-3.0-or-later",
  "devDependencies": {
    "typescript": "^4.0.5"
  },
  "scripts": {
    "build": "tsc"
  },
  "private": true
}
