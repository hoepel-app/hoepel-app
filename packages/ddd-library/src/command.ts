import { Message } from './message'

export type CommandRequestedBy = {
  type: 'user'
  uid: string
  email: string
}

export type CommandRejected = { status: 'rejected' }
export type CommandAccepted = { status: 'accepted' }
export type CommandResult = CommandAccepted | CommandRejected

export type Command<T> = {
  /** @deprecated */
  toProps(): T

  requestedBy: CommandRequestedBy
} & Message<T>
