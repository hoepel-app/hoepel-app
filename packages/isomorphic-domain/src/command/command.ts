export type CommandRejected = { status: 'rejected' }
export type CommandAccepted = { status: 'accepted' }
export type CommandResult = CommandAccepted | CommandRejected

export type Command<T> = {
  toProps(): T
}
