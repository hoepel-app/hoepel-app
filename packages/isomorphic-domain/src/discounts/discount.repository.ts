import { Observable } from 'rxjs'
import { Discounts } from './discounts'

export type DiscountsRepository = {
  getForTenant(tenantId: string): Observable<Discounts>
  put(entity: Discounts): Promise<void>
}
