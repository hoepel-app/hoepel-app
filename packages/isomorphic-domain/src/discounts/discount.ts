import { Price } from '@hoepel.app/types'

type AbsoluteDiscountValue = {
  kind: 'relative-discount'
  discountPercent: number
}
type RelativeDiscountValue = {
  kind: 'absolute-discount'
  discountCents: number
}
type DiscountValue = AbsoluteDiscountValue | RelativeDiscountValue

export type DiscountProps = {
  name: string
  discount: DiscountValue
}

export class Discount {
  private constructor(private readonly props: DiscountProps) {}

  toProps(): DiscountProps {
    return this.props
  }

  static fromProps(props: DiscountProps): Discount {
    return new Discount(props)
  }

  static createRelativeDiscount(name: string, percent: number): Discount {
    const discountPercent = Math.round(Math.min(100, Math.max(percent, 0)))

    return new Discount({
      name,
      discount: {
        kind: 'relative-discount',
        discountPercent: isNaN(discountPercent) ? 0 : discountPercent,
      },
    })
  }

  static createAbsoluteDiscount(name: string, price: Price): Discount {
    return new Discount({
      name,
      discount: {
        kind: 'absolute-discount',
        discountCents: price.totalCents,
      },
    })
  }

  /**
   * Apply a list of discounts to a price. The discounts will be applied in the array order.
   *
   * The result will never be negative (always at least 0.00EUR)
   *
   * @param to The price to apply these discounts to
   * @param discounts The list of discounts to be applied
   */
  static applyAllDiscounts(to: Price, discounts: readonly Discount[]): Price {
    return discounts.reduce((acc, current) => {
      return current.applyTo(acc)
    }, to)
  }

  /**
   * Applies this discount to a price.
   * The result will never be negative (always at least 0.00EUR)
   */
  applyTo(price: Price): Price {
    switch (this.props.discount.kind) {
      case 'absolute-discount': {
        return price.subtract(this.absoluteDiscount)
      }
      case 'relative-discount': {
        const pct = this.discountPercentage * 0.01
        return price.multiply(1 - pct)
      }
    }
  }

  get isRelativeDiscount(): boolean {
    return this.props.discount.kind === 'relative-discount'
  }

  get isAbsoluteDiscount(): boolean {
    return this.props.discount.kind === 'absolute-discount'
  }

  /**
   * Return the discount as a percentage ([0,100]).
   * If this discount is an absolute discount, returns 0%
   */
  get discountPercentage(): number {
    switch (this.props.discount.kind) {
      case 'absolute-discount':
        return 0
      case 'relative-discount':
        return this.props.discount.discountPercent
    }
  }

  /**
   * Return the absolute discount as a price to be substracted
   * If this discount is a relative discount, returns 0.00EUR
   */
  get absoluteDiscount(): Price {
    switch (this.props.discount.kind) {
      case 'absolute-discount':
        return Price.fromCents(this.props.discount.discountCents)
      case 'relative-discount':
        return Price.zero
    }
  }

  get name(): string {
    return this.props.name
  }

  withName(name: string): Discount {
    return Discount.fromProps({
      ...this.toProps(),
      name,
    })
  }
}
