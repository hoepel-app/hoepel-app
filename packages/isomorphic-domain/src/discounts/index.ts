export { AddDiscountCommand } from './commands/add-discount.command'
export { RemoveDiscountCommand } from './commands/remove-discount.command'
export { RenameDiscountCommand } from './commands/rename-discount.command'
export { DiscountsRepository } from './discount.repository'
export { Discounts, DiscountsProps } from './discounts'
export { Discount, DiscountProps } from './discount'
export { DiscountsApplicationService } from './discounts.application-service'
