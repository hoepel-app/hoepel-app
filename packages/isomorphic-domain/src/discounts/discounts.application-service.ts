import { DiscountsRepository } from './discount.repository'
import { Discounts } from './discounts'
import { Observable } from 'rxjs'
import { AddDiscountCommand } from './commands/add-discount.command'
import { CommandResult } from '@hoepel.app/ddd-library'
import { first } from 'rxjs/operators'
import { RenameDiscountCommand } from './commands/rename-discount.command'
import { RemoveDiscountCommand } from './commands/remove-discount.command'

export class DiscountsApplicationService {
  constructor(private readonly repo: DiscountsRepository) {}

  findDiscounts(tenantId: string): Observable<Discounts> {
    return this.repo.getForTenant(tenantId)
  }

  async addDiscount(command: AddDiscountCommand): Promise<CommandResult> {
    const discounts = await this.repo
      .getForTenant(command.tenantId)
      .pipe(first())
      .toPromise()

    if (command.discount.name.trim() === '') {
      return {
        status: 'rejected',
        reason: 'Discount name is empty',
      }
    }

    if (!discounts.mayAddDiscount(command.discount)) {
      return {
        status: 'rejected',
        reason: `May not add the discount with name ${command.discount.name}`,
      }
    }

    await this.repo.put(discounts.withDiscountAdded(command.discount))

    return { status: 'accepted' }
  }

  async renameDiscount(command: RenameDiscountCommand): Promise<CommandResult> {
    const discounts = await this.repo
      .getForTenant(command.tenantId)
      .pipe(first())
      .toPromise()

    if (command.newName.trim() === '') {
      return {
        status: 'rejected',
        reason: `New name is empty`,
      }
    }

    if (!discounts.discountWithNameExists(command.oldName)) {
      return {
        status: 'rejected',
        reason: `A discount with name ${command.oldName} does not exist`,
      }
    }

    if (discounts.discountWithNameExists(command.newName)) {
      return {
        status: 'rejected',
        reason: `A discount with name ${command.newName} already exists`,
      }
    }

    await this.repo.put(
      discounts.withDiscountRenamed(command.oldName, command.newName)
    )

    return { status: 'accepted' }
  }

  async removeDiscount(command: RemoveDiscountCommand): Promise<CommandResult> {
    const discounts = await this.repo
      .getForTenant(command.tenantId)
      .pipe(first())
      .toPromise()

    if (!discounts.discountWithNameExists(command.discountName)) {
      return {
        status: 'rejected',
        reason: `A discount with name ${command.discountName} does not exist`,
      }
    }

    await this.repo.put(discounts.withDiscountRemoved(command.discountName))

    return { status: 'accepted' }
  }
}
