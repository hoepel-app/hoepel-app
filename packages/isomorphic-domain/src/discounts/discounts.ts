import { Aggregate } from '@hoepel.app/ddd-library'
import { DiscountProps, Discount } from './discount'

export type DiscountsProps = {
  readonly tenantId: string
  readonly discounts: DiscountProps[]
}

export class Discounts implements Aggregate {
  private constructor(private readonly props: DiscountsProps) {}

  static createEmpty(tenantId: string): Discounts {
    return new Discounts({ tenantId, discounts: [] })
  }

  get id(): string {
    return `${this.tenantId}-discounts`
  }

  get tenantId(): string {
    return this.props.tenantId
  }

  get names(): ReadonlySet<string> {
    return new Set(this.discounts.map((discount) => discount.name))
  }

  get namesSorted(): readonly string[] {
    return [...this.names].sort()
  }

  get discounts(): readonly Discount[] {
    return [
      ...this.props.discounts.map((props) => Discount.fromProps(props)),
    ].sort((a, b) => a.name.localeCompare(b.name))
  }

  toProps(): DiscountsProps {
    return this.props
  }

  static fromProps(props: DiscountsProps): Discounts {
    return new Discounts(props)
  }

  get isEmpty(): boolean {
    return this.props.discounts.length === 0
  }

  mayAddDiscount(discount: Discount): boolean {
    return !this.discountWithNameExists(discount.name)
  }

  discountWithNameExists(name: string): boolean {
    return this.names.has(name)
  }

  findDiscountWithName(name: string): Discount | null {
    return this.discounts.find((discount) => discount.name === name) ?? null
  }

  withDiscountAdded(discount: Discount): Discounts {
    if (!this.mayAddDiscount(discount)) {
      return this
    }

    return Discounts.fromProps({
      ...this.toProps(),
      discounts: [
        ...this.discounts.map((d) => d.toProps()),
        discount.toProps(),
      ],
    })
  }

  withDiscountRenamed(oldName: string, newName: string): Discounts {
    const oldDiscount = this.findDiscountWithName(oldName)

    if (oldDiscount == null || this.discountWithNameExists(newName)) {
      return this
    }

    const renamedDiscount = oldDiscount.withName(newName)

    return Discounts.fromProps({
      ...this.toProps(),
      discounts: [
        ...this.discounts
          .filter((d) => d.name !== oldName)
          .map((d) => d.toProps()),
        renamedDiscount.toProps(),
      ],
    })
  }

  withDiscountRemoved(name: string): Discounts {
    return Discounts.fromProps({
      ...this.toProps(),
      discounts: [
        ...this.discounts
          .filter((d) => d.name !== name)
          .map((d) => d.toProps()),
      ],
    })
  }
}
