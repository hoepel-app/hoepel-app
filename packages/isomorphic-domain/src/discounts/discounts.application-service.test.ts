import { CommandMetadata } from '@hoepel.app/ddd-library'
import { Discounts } from './discounts'
import { Price } from '@hoepel.app/types'
import { Discount } from './discount'
import { DiscountsApplicationService } from './discounts.application-service'
import { AddDiscountCommand } from './commands/add-discount.command'
import { of } from 'rxjs'

import '@hoepel.app/ddd-library-test-utils'
import { RenameDiscountCommand } from './commands/rename-discount.command'
import { RemoveDiscountCommand } from './commands/remove-discount.command'

describe('DiscountsApplicationService', () => {
  const exampleDiscounts = (tenantId: string): Discounts =>
    Discounts.createEmpty(tenantId)
      .withDiscountAdded(
        Discount.createAbsoluteDiscount('1 euro korting', Price.fromCents(100))
      )
      .withDiscountAdded(Discount.createRelativeDiscount('Kansentarief', 60))
      .withDiscountAdded(Discount.createRelativeDiscount('Tweede kind', 50))

  const commandMetadata: CommandMetadata = {
    commandId: 'my-command-id-123',
    requestedBy: {
      email: 'test@example.org',
      type: 'user',
      uid: 'my-uid-123',
    },
    tenantId: 'some-tenant-id',
    timestamp: new Date('2020-04-25T19:43:29.161Z'),
  }

  describe('addDiscount', () => {
    it('rejects when adding a discount with a name that is already taken', async () => {
      const repo = {
        put: jest.fn(),
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleDiscounts(tenantId))
        ),
      }

      const service = new DiscountsApplicationService(repo)
      const command = AddDiscountCommand.create(
        Discount.createAbsoluteDiscount('Kansentarief', Price.fromCents(100)),
        commandMetadata
      )

      const result = await service.addDiscount(command)

      expect(result).toBeRejectedWithReason(
        'May not add the discount with name Kansentarief'
      )
      expect(repo.put).not.toBeCalled()
    })

    it('rejects when adding a discount with an empty name', async () => {
      const repo = {
        put: jest.fn(),
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleDiscounts(tenantId))
        ),
      }

      const service = new DiscountsApplicationService(repo)
      const command = AddDiscountCommand.create(
        Discount.createAbsoluteDiscount('', Price.fromCents(100)),
        commandMetadata
      )

      const result = await service.addDiscount(command)

      expect(result).toBeRejectedWithReason('Discount name is empty')
      expect(repo.put).not.toBeCalled()
    })

    it('can add a discount', async () => {
      const repo = {
        put: jest.fn(),
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleDiscounts(tenantId))
        ),
      }

      const service = new DiscountsApplicationService(repo)
      const command = AddDiscountCommand.create(
        Discount.createAbsoluteDiscount('Test', Price.fromCents(100)),
        commandMetadata
      )

      const result = await service.addDiscount(command)

      expect(result).toBeAccepted()
      expect(repo.put).toBeCalledTimes(1)
      expect(repo.put.mock.calls).toMatchSnapshot()
    })
  })

  describe('renameDiscount', () => {
    it('rejects empty new name', async () => {
      const repo = {
        put: jest.fn(),
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleDiscounts(tenantId))
        ),
      }

      const service = new DiscountsApplicationService(repo)
      const command = RenameDiscountCommand.create(
        'Tweede kind',
        '',
        commandMetadata
      )

      const result = await service.renameDiscount(command)

      expect(result).toBeRejectedWithReason('New name is empty')
      expect(repo.put).not.toBeCalled()
    })

    it('rejects renaming a discount that does not exist', async () => {
      const repo = {
        put: jest.fn(),
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleDiscounts(tenantId))
        ),
      }

      const service = new DiscountsApplicationService(repo)
      const command = RenameDiscountCommand.create(
        'Derde kind',
        'Naam',
        commandMetadata
      )

      const result = await service.renameDiscount(command)

      expect(result).toBeRejectedWithReason(
        'A discount with name Derde kind does not exist'
      )
      expect(repo.put).not.toBeCalled()
    })

    it('rejects renaming when the new name already exists', async () => {
      const repo = {
        put: jest.fn(),
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleDiscounts(tenantId))
        ),
      }

      const service = new DiscountsApplicationService(repo)
      const command = RenameDiscountCommand.create(
        'Tweede kind',
        'Kansentarief',
        commandMetadata
      )

      const result = await service.renameDiscount(command)

      expect(result).toBeRejectedWithReason(
        'A discount with name Kansentarief already exists'
      )
      expect(repo.put).not.toBeCalled()
    })

    it('renames a discount', async () => {
      const repo = {
        put: jest.fn(),
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleDiscounts(tenantId))
        ),
      }

      const service = new DiscountsApplicationService(repo)
      const command = RenameDiscountCommand.create(
        'Tweede kind',
        'Nieuwe naam',
        commandMetadata
      )

      const result = await service.renameDiscount(command)

      expect(result).toBeAccepted()
      expect(repo.put).toBeCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })

  describe('removeDiscount', () => {
    it('rejects when name not found', async () => {
      const repo = {
        put: jest.fn(),
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleDiscounts(tenantId))
        ),
      }

      const service = new DiscountsApplicationService(repo)
      const command = RemoveDiscountCommand.create('Something', commandMetadata)

      const result = await service.removeDiscount(command)

      expect(result).toBeRejectedWithReason(
        'A discount with name Something does not exist'
      )
      expect(repo.put).not.toBeCalled()
    })

    it('removes a discount', async () => {
      const repo = {
        put: jest.fn(),
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleDiscounts(tenantId))
        ),
      }

      const service = new DiscountsApplicationService(repo)
      const command = RemoveDiscountCommand.create(
        'Tweede kind',
        commandMetadata
      )

      const result = await service.removeDiscount(command)

      expect(result).toBeAccepted()
      expect(repo.put).toBeCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })
})
