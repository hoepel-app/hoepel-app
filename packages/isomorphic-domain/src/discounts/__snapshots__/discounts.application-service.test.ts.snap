// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscountsApplicationService addDiscount can add a discount 1`] = `
Array [
  Array [
    Discounts {
      "props": Object {
        "discounts": Array [
          Object {
            "discount": Object {
              "discountCents": 100,
              "kind": "absolute-discount",
            },
            "name": "1 euro korting",
          },
          Object {
            "discount": Object {
              "discountPercent": 60,
              "kind": "relative-discount",
            },
            "name": "Kansentarief",
          },
          Object {
            "discount": Object {
              "discountPercent": 50,
              "kind": "relative-discount",
            },
            "name": "Tweede kind",
          },
          Object {
            "discount": Object {
              "discountCents": 100,
              "kind": "absolute-discount",
            },
            "name": "Test",
          },
        ],
        "tenantId": "some-tenant-id",
      },
    },
  ],
]
`;

exports[`DiscountsApplicationService removeDiscount removes a discount 1`] = `
Array [
  Discounts {
    "props": Object {
      "discounts": Array [
        Object {
          "discount": Object {
            "discountCents": 100,
            "kind": "absolute-discount",
          },
          "name": "1 euro korting",
        },
        Object {
          "discount": Object {
            "discountPercent": 60,
            "kind": "relative-discount",
          },
          "name": "Kansentarief",
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;

exports[`DiscountsApplicationService renameDiscount renames a discount 1`] = `
Array [
  Discounts {
    "props": Object {
      "discounts": Array [
        Object {
          "discount": Object {
            "discountCents": 100,
            "kind": "absolute-discount",
          },
          "name": "1 euro korting",
        },
        Object {
          "discount": Object {
            "discountPercent": 60,
            "kind": "relative-discount",
          },
          "name": "Kansentarief",
        },
        Object {
          "discount": Object {
            "discountPercent": 50,
            "kind": "relative-discount",
          },
          "name": "Nieuwe naam",
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;
